functorch/csrc/init.cpp

```diff
@@ -29,10 +30,56 @@ static bool has_level(const Tensor& self, int64_t level) {
   return batched->level() >= level;
 }
 
+static bool has_functional_level(const Tensor& self, int64_t level) {
+  const auto* functional = dynamic_cast<FunctionalTensorWrapper*>(self.unsafeGetTensorImpl());
+  if (!functional) {
+    return false;
+  }
+  return functional->level() >= level;
+}
+
 Tensor _add_batch_dim(const Tensor& self, int64_t batch_dim, int64_t level) {
   return addBatchDim(self, batch_dim, level);
 }
 
+Tensor _wrap_functional_tensor(const Tensor& self, int64_t level) {
+  auto t = at::functionalization::impl::to_functional_tensor(self);
+  at::functionalization::impl::unsafeGetFunctionalWrapper(t)->set_level(level);
+  return t;
+}
+
+void _assert_wrapped_functional(const Tensor& unwrapped, const Tensor& wrapped) {
+  TORCH_INTERNAL_ASSERT(at::functionalization::impl::isFunctionalTensor(wrapped));
+  TORCH_INTERNAL_ASSERT(!at::functionalization::impl::isFunctionalTensor(unwrapped));
+  auto wrapped_impl = at::functionalization::impl::unsafeGetFunctionalWrapper(wrapped);
+  auto& wrapped_inner = wrapped_impl->value();
+  TORCH_INTERNAL_ASSERT(unwrapped.unsafeGetTensorImpl() == wrapped_inner.unsafeGetTensorImpl())
+}
+
+void _propagate_functional_input_mutation(const Tensor& unwrapped, const Tensor& wrapped) {
+  TORCH_INTERNAL_ASSERT(at::functionalization::impl::isFunctionalTensor(wrapped));
+  TORCH_INTERNAL_ASSERT(!at::functionalization::impl::isFunctionalTensor(unwrapped));
+  auto wrapped_impl = at::functionalization::impl::unsafeGetFunctionalWrapper(wrapped);
+  // Ensure that the input is up to date by committing any pending updates to the alias.
+  wrapped_impl->apply_updates();
+  wrapped_impl->regenerate_from_base();
+  auto& wrapped_inner = wrapped_impl->value();
+  // It would probably be more reasonable to check that the two tensors are aliased,
+  // but we can't do that unless we give BatchedTensorImpl a notion of storage.
+  if (unwrapped.unsafeGetTensorImpl() == wrapped_inner.unsafeGetTensorImpl()) {
+  } else {
+      TORCH_INTERNAL_ASSERT(unwrapped.nbytes() == wrapped_inner.nbytes());
+      if (unwrapped.sizes() == wrapped_inner.sizes()) {
+          std::cout << "propagating changes to a mutated input!" << std::endl;
+          unwrapped.copy_(wrapped_inner);
+      } else {
+          std::cout << "An inplace-mutation op (like transpose_() was called on an input to the functionalization pass.";
+          std::cout << " Propagating those mutations to the input is currently not supported." << std::endl;
```

Comment:
Should this be an error message? Or a warning?

Reply:
I'm thinking of making it an error, but let me know if you have an opinion. Hopefully we can remove `transpose_()` et al in the near future anyway.